server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -643,14 +641,19 @@ private SearchPhaseResult executeQueryPhase(ShardSearchRequest request, SearchSh
                 tracer.stopTrace();
             }
             if (request.numberOfShards() == 1) {
+                // we already have query results, but we can run fetch at the same time
                 context.addFetchResult();
+                // This will return a QueryFetchSearchResult, which incRef's the QuerySearchResult it holds
                 return executeFetchPhase(readerContext, context, afterQueryTime);
             } else {
                 // Pass the rescoreDocIds to the queryResult to send them the coordinating node and receive them back in the fetch phase.
                 // We also pass the rescoreDocIds to the LegacyReaderContext in case the search state needs to stay in the data node.
                 final RescoreDocIds rescoreDocIds = context.rescoreDocIds();
                 context.queryResult().setRescoreDocIds(rescoreDocIds);
                 readerContext.setRescoreDocIds(rescoreDocIds);
+                // Since we are returning the raw QuerySearchResult here, there's no opportunity for the object that will be holding that
+                // reference to incRef it before our try block closes and decRef's it when it closes the search context
+                context.queryResult().incRef();
```

Comment:
It's common, and totally reasonable to assume (require?) that the caller will take ownership of the returned value, and release it when no longer needed. I think it's not clear from the code that this is required, so it might be worth documenting it in this method's JavaDocs. I would recommend against the inline comments here however, stuff like this seems useful when you write it but tends to just become clutter over time.

I think the caller does behave as expected in this case, right? It looks like we pass it (pretty) directly to a `ChannelActionListener` which does `decRef()` once it no longer needs it. (aside: we really should document this)

There's definitely a risk of leaks with this kind of API tho, so it's important that a leak in this area would be caught by tests. Is that the case here? For instance, if it's using potentially-recycled pages then we should be using a `CountingPageCacheRecycler` (preferred) or `MockPageCacheRecycler` (relies on GC).

Reply:
Yes, the caller does the right thing and the tests do catch it if we don't.  Honestly, the testing around this has been very good so far.

And you're right, I should just write javadoc for these methods.  I hadn't because I've really only been focusing on this one aspect, which is not the main thrust of what's going on here at all, but they currently have none at all, so I'm sure I can do better than that.